### ISSUE TITLE
(asan) heap-buffer-overflow in screen_del_lines()

### ISSUE THREAD

**Issue by dpelle** [vim/vim#4621]
I have a script that builds vim and runs tests with various configure options.
While running it, asan found a heap-buffer-overflow bug.  Vim-8.1.1639
was configured with:

```
$ CC=clang-8 ./configure --with-features=huge --enable-gui=gtk3 --disable-rightleft
```

Bug found with asan while running tests:
```
VIMRUNTIME=../../runtime; export VIMRUNTIME;  ../vim -f  -u unix.vim -u NONE -U NONE --noplugin --not-a-term -S runtest.vim test_gui.vim
[?1049h[22;0;0t[?1h=[?2004h[1;55r[?12h[?12l[22;2t[22;1t[27m[23m[29m[m[H[2J[?25l[55;1H"test_gui.vim" 821 lines, 23235 characters

Executing Test_1_set_secure()[?2004l[?25h[?2004l[?1l>[?1049l[23;0;0t=================================================================
==5173==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x611000254288 at pc 0x000000ac668a bp 0x7fff47358c50 sp 0x7fff47358c48
READ of size 4 at 0x611000254288 thread T0
    #0 0xac6689 in screen_del_lines /home/dope/sb/vim/src/screen.c:10358:32
    #1 0xad2fd2 in win_do_lines /home/dope/sb/vim/src/screen.c:9917:15
    #2 0xad3654 in win_del_lines /home/dope/sb/vim/src/screen.c:9818:14
    #3 0xaafc37 in win_update /home/dope/sb/vim/src/screen.c:1504:11
    #4 0xab7eb9 in win_update /home/dope/sb/vim/src/screen.c:2347:3
    #5 0xaa44ce in update_screen /home/dope/sb/vim/src/screen.c:785:6
    #6 0x6e3f99 in ex_redraw /home/dope/sb/vim/src/ex_docmd.c:8358:5
    #7 0x6c2c08 in do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2500:2
    #8 0x6b53dc in do_cmdline /home/dope/sb/vim/src/ex_docmd.c:995:17
    #9 0xce64b3 in call_user_func /home/dope/sb/vim/src/userfunc.c:1063:2
    #10 0xce17e8 in call_func /home/dope/sb/vim/src/userfunc.c:1621:7
    #11 0xce01a8 in get_func_tv /home/dope/sb/vim/src/userfunc.c:490:8
    #12 0xcf72cd in ex_call /home/dope/sb/vim/src/userfunc.c:3345:6
    #13 0x6c2c08 in do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2500:2
    #14 0x6b53dc in do_cmdline /home/dope/sb/vim/src/ex_docmd.c:995:17
    #15 0x5df6f2 in ex_execute /home/dope/sb/vim/src/eval.c:8925:6
    #16 0x6c2c08 in do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2500:2
    #17 0x6b53dc in do_cmdline /home/dope/sb/vim/src/ex_docmd.c:995:17
    #18 0xce64b3 in call_user_func /home/dope/sb/vim/src/userfunc.c:1063:2
    #19 0xce17e8 in call_func /home/dope/sb/vim/src/userfunc.c:1621:7
    #20 0xce01a8 in get_func_tv /home/dope/sb/vim/src/userfunc.c:490:8
    #21 0xcf72cd in ex_call /home/dope/sb/vim/src/userfunc.c:3345:6
    #22 0x6c2c08 in do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2500:2
    #23 0x6b53dc in do_cmdline /home/dope/sb/vim/src/ex_docmd.c:995:17
    #24 0x6aa0f7 in do_source /home/dope/sb/vim/src/ex_cmds2.c:3638:5
    #25 0x6a7e4e in cmd_source /home/dope/sb/vim/src/ex_cmds2.c:3257:14
    #26 0x6a7fec in ex_source /home/dope/sb/vim/src/ex_cmds2.c:3232:2
    #27 0x6c2c08 in do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2500:2
    #28 0x6b53dc in do_cmdline /home/dope/sb/vim/src/ex_docmd.c:995:17
    #29 0x6b8ce3 in do_cmdline_cmd /home/dope/sb/vim/src/ex_docmd.c:596:12
    #30 0xe59fc8 in exe_commands /home/dope/sb/vim/src/main.c:2975:2
    #31 0xe5699c in vim_main2 /home/dope/sb/vim/src/main.c:797:2
    #32 0xe4e933 in main /home/dope/sb/vim/src/main.c:444:12
    #33 0x7fc71a2e4b96 in __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:310
    #34 0x428379 in _start (/home/dope/sb/vim/src/vim+0x428379)

0x611000254288 is located 0 bytes to the right of 200-byte region [0x6110002541c0,0x611000254288)
allocated by thread T0 here:
    #0 0x4d3d23 in __interceptor_malloc (/home/dope/sb/vim/src/vim+0x4d3d23)
    #1 0x88b24b in lalloc /home/dope/sb/vim/src/misc2.c:924:11
    #2 0xacd01d in screenalloc /home/dope/sb/vim/src/screen.c:9034:22
    #3 0xaaaa71 in screenclear /home/dope/sb/vim/src/screen.c:9282:5
    #4 0xc4c8fa in set_shellsize /home/dope/sb/vim/src/term.c:3462:2
    #5 0xc512d0 in shell_resized /home/dope/sb/vim/src/term.c:3376:5
    #6 0xd3edf5 in gui_resize_shell /home/dope/sb/vim/src/gui.c:1537:2
    #7 0xd748bf in form_configure_event /home/dope/sb/vim/src/gui_gtk_x11.c:4180:5
    #8 0x7fc71de967fa  (/usr/lib/x86_64-linux-gnu/libgtk-3.so.0+0x2317fa)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/dope/sb/vim/src/screen.c:10358:32 in screen_del_lines
Shadow bytes around the buggy address:
  0x0c2280042800: fd fd fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2280042810: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2280042820: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2280042830: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c2280042840: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c2280042850: 00[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2280042860: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2280042870: fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa fa
  0x0c2280042880: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c2280042890: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c22800428a0: fd fd fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==5173==ABORTING
Vim: Caught deadly signal ABRT
Vim: Finished.
Aborted (core dumped)
Makefile:168: recipe for target 'test_gui.res' failed
make[2]: *** [test_gui.res] Error 134
make[2]: Leaving directory '/home/dope/sb/vim/src/testdir'
Makefile:2143: recipe for target 'scripttests' failed
make[1]: *** [scripttests] Error 2
make[1]: Leaving directory '/home/dope/sb/vim/src'
Makefile:36: recipe for target 'test' failed
make: *** [test] Error 2
```

Maybe a race condition if gui_resize_shell() is
called asynchronously?

I'll check later whether it's reproducible all the time.

**Environment**
 - Vim 8.1.1639 (huge, gtk3, --disable-rightleft)
 - xubuntu-18.04.2, x86_64

**Comment by brammool** [comment]
Dominique wrote:

> I have a script that builds vim and runs tests with various configure
> options.  While running it, asan found a heap-buffer-overflow bug.
> Vim-8.1.1639 was configured with:
> 
> ```
> $ CC=clang-8 ./configure --with-features=huge --enable-gui=gtk3 --disable-rightleft
> ```
> 
> Bug found with asan while running tests:
> ```
> VIMRUNTIME=../../runtime; export VIMRUNTIME;  ../vim -f  -u unix.vim -u NONE -U NONE --noplugin --not-a-term -S runtest.vim test_gui.vim
> [?1049h[22;0;0t[?1h=[?2004h[1;55r[?12h[?12l[22;2t[22;1t[27m[23m[29m[m[H[2J[?25l[55;1H"test_gui.vim" 821 lines, 23235 characters
> 
> Executing Test_1_set_secure()[?2004l[?25h[?2004l[?1l>[?1049l[23;0;0t=================================================================
> ==5173==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x611000254288 at pc 0x000000ac668a bp 0x7fff47358c50 sp 0x7fff47358c48
> READ of size 4 at 0x611000254288 thread T0
>     #0 0xac6689 in screen_del_lines /home/dope/sb/vim/src/screen.c:10358:32
>     #1 0xad2fd2 in win_do_lines /home/dope/sb/vim/src/screen.c:9917:15
>     #2 0xad3654 in win_del_lines /home/dope/sb/vim/src/screen.c:9818:14
>     #3 0xaafc37 in win_update /home/dope/sb/vim/src/screen.c:1504:11
>     #4 0xab7eb9 in win_update /home/dope/sb/vim/src/screen.c:2347:3
>     #5 0xaa44ce in update_screen /home/dope/sb/vim/src/screen.c:785:6
>     #6 0x6e3f99 in ex_redraw /home/dope/sb/vim/src/ex_docmd.c:8358:5

At the start of update_screen() thre is a call to screen_valid(), thus
LineOffset[] should have the right size.  So either somehow screen_Rows
and Rows differ, or the array index is actually wrong here.
Oh, there is another call to screen_valid() at the top of
screen_del_lines(), thus that must be OK.  If it did resize there,
perhaps the arguments to screen_del_lines() are now wrong though.

> 0x611000254288 is located 0 bytes to the right of 200-byte region [0x6110002541c0,0x611000254288)
> allocated by thread T0 here:
>     #0 0x4d3d23 in __interceptor_malloc (/home/dope/sb/vim/src/vim+0x4d3d23)
>     #1 0x88b24b in lalloc /home/dope/sb/vim/src/misc2.c:924:11
>     #2 0xacd01d in screenalloc /home/dope/sb/vim/src/screen.c:9034:22

That's allocating LineOffset[].

> Maybe a race condition if gui_resize_shell() is
> called asynchronously?
> 
> I'll check later whether it's reproducible all the time.

It's hard to do anything with more information.  If you can reproduce,
adding some ch_log() calls might reveal the problem.

Hmm, If you can reproduce it, see if this helps:


--- ../../git/vim81/src/term.c	2019-06-06 15:19:28.251404512 +0200
+++ term.c	2019-07-06 13:55:50.301056261 +0200
@@ -3423,9 +3423,9 @@
     if (width < 0 || height < 0)    /* just checking... */
 	return;
 
-    if (State == HITRETURN || State == SETWSIZE)
+    if (State == HITRETURN || State == SETWSIZE || updating_screen)
     {
-	/* postpone the resizing */
+	// postpone the resizing
 	State = SETWSIZE;
 	return;
     }

-- 
Shit makes the flowers grow and that's beautiful

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
I suppose this is fixed by 847a5d69a826c02a7cb8a14884f3fef69d74fed7, so closing.
